File: create_secret.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import os
import asyncio
from google.cloud import secretmanager_v1 as secretmanager
from google.api_core.exceptions import GoogleAPICallError

async def get_secret(project_id: str, secret_id: str, location: str = None) -> secretmanager.Secret:
    """
    Gets a secret from Google Cloud Secret Manager.

    Args:
        project_id: The ID of the Google Cloud project.
        secret_id: The ID of the secret to retrieve.
        location: (Optional) The location of the secret. If not provided,
                  the global secret path will be used.

    Returns:
        The retrieved Secret object.
    """
    # [START secretmanager_v1_secretmanagerservice_secret_create_async]
    client = secretmanager.SecretManagerServiceAsyncClient()

    if location:
        name = client.secret_path(project_id, location, secret_id)
    else:
        name = client.secret_path(project_id, secret_id)

    request = secretmanager.GetSecretRequest(name=name)

    response = None
    try:
        response = await client.get_secret(request=request)
    except GoogleAPICallError as e:
        print(f"Error getting secret '{secret_id}': {e}")
        raise
    finally:
        await client.close()

    return response
    # [END secretmanager_v1_secretmanagerservice_secret_create_async]

async def main():
    parser = argparse.ArgumentParser(
        description="Get a secret from Google Cloud Secret Manager."
    )
    parser.add_argument(
        "--secret-id",
        type=str,
        required=True,
        help="The ID of the secret to retrieve.",
    )
    parser.add_argument(
        "--location",
        type=str,
        required=False,
        help="The location of the secret (e.g., 'global' or 'us-central1').",
    )

    args = parser.parse_args()

    project_id = os.environ.get("GCP_PROJECT_ID")
    if not project_id:
        raise ValueError("GCP_PROJECT_ID environment variable not set.")

    try:
        secret = await get_secret(project_id, args.secret_id, args.location)
        print(secret)
    except Exception as e:
        print(f"Failed to retrieve secret: {e}")
        exit(1)

if __name__ == "__main__":
    asyncio.run(main())

File: createSecret.js
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

"use strict";

const { SecretManagerServiceClient } = require("@google-cloud/secret-manager");

/**
 * Gets metadata for a given Secret.
 *
 * @param {string} projectId The Google Cloud Project ID.
 * @param {string} secretId The ID of the secret to retrieve.
 * @returns {Promise<object>} The Secret metadata.
 */
async function getSecret(projectId, secretId) {
  // [START secretmanager_v1_secretmanagerservice_get_secret_async]
  // Initialize the Secret Manager client.
  // The client can be reused across multiple calls to reduce overhead.
  const client = new SecretManagerServiceClient();

  // Construct the resource name for the secret.
  // The format is projects/{project}/secrets/{secret}
  // For regional secrets, it would be projects/{project}/locations/{location}/secrets/{secret}
  // This sample uses the global secret format.
  const name = `projects/${projectId}/secrets/${secretId}`;

  try {
    // Call the getSecret API to retrieve the secret metadata.
    const [secret] = await client.getSecret({ name });
    return secret;
  } catch (err) {
    // Re-throw the error to be caught by the main function's try/catch block.
    throw new Error(`Failed to get secret '${name}': ${err.message || err}`);
  } finally {
    // Close the client to release resources.
    // This is important for long-running processes.
    await client.close();
  }
  // [END secretmanager_v1_secretmanagerservice_get_secret_async]
}

/**
 * Main function to parse arguments, call the getSecret function, and print the result.
 */
async function main() {
  try {
    // Manually parse command-line arguments.
    const args = {};
    process.argv.slice(2).forEach((arg) => {
      if (arg.startsWith("--")) {
        const parts = arg.substring(2).split("=");
        if (parts.length === 2) {
          args[parts[0]] = parts[1];
        } else {
          args[parts[0]] = true;
        }
      }
    });

    const secretId = args["secret-id"];
    if (!secretId) {
      console.error("Error: --secret-id argument is required.");
      console.error("Usage: node get_secret.js --secret-id=<your-secret-id>");
      process.exit(1);
    }

    // Get the Google Cloud Project ID from environment variables.
    const projectId = process.env.GCP_PROJECT_ID;
    if (!projectId) {
      throw new Error(
        "GCP_PROJECT_ID environment variable is not set. Please set it to your Google Cloud Project ID.",
      );
    }

    // Call the primary sample function.
    const secret = await getSecret(projectId, secretId);

    // Print the result to standard output in JSON format.
    console.log(JSON.stringify(secret, null, 2));
  } catch (error) {
    // Log any errors and set a non-zero exit code.
    console.error("Error:", error.message || error);
    process.exitCode = 1;
  }
}

// Execute the main function when the script is run directly.
if (require.main === module) {
  main();
}
